tests/testthat/test-indentation_linter.R

```diff
@@ -0,0 +1,423 @@
+      abc[[
+        'elem'
+      ]]
+
+      def[[a,
+           b]]
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("indentation works with control flow statements", {
+  linter <- indentation_linter()
+
+  expect_lint(
+    trim_some("
+      if (TRUE) {
+        do_something
+      } else {
+        do_other_thing
+      }
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      while (1 > 2) {
+      do_something
+      }
+    "),
+    "Indentation",
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      if (FALSE) {
+        do_something
+        } else {
+        do_other_thing
+      }
+    "),
+    "Indentation",
+    linter
+  )
+})
+
+test_that("indentation within string constants is ignored", {
+  expect_lint(
+    trim_some("
+      x <- '
+        an indented string
+      '
+    "),
+    NULL,
+    indentation_linter()
+  )
+})
+
+test_that("combined hanging and block indent works", {
+  linter <- indentation_linter()
+  expect_lint(
+    trim_some("
+      func(hang, and,
+           block(
+             combined
+           ))
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      func(ha,
+           func2(ab,
+                 block(
+                   indented
+                 )))
+    "),
+    NULL,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      func(func2(
+        a = 42
+      ))
+    "),
+    NULL,
+    linter
+  )
+
+  # Adapted from cli R/ansi.R L231-234
+  expect_lint(
+    trim_some("
+      stopifnot(is.character(style) && length(style) == 1 ||
+                  is_rgb_matrix(style) && ncol(style) == 1,
+                is.logical(bg) && length(bg) == 1,
+                is.numeric(colors) && length(colors) == 1)
+    "),
+    NULL,
+    linter
+  )
+
+  # Adapted from cli inst/scripts/up.R L26-37
+  expect_lint(
+    trim_some("
+      http_head(url, ...)$
+        then(function(res) {
+          if (res$status_code < 300) {
+            cli_alert_success()
+          } else {
+            cli_alert_danger()
+          }
+        })$
+        catch(error = function(err) {
+          e <- if (grepl('timed out', err$message)) 'timed out' else 'error'
+          cli_alert_danger()
+        })
+    "),
+    NULL,
+    linter
+  )
+})
+
+test_that("use_hybrid_indent works", {
+  code_hybrid <- "map(x, f,\n  extra_arg = 42\n)"
+  code_non_hybrid <- "map(x, f,\n    extra_arg = 42\n)"
+  code_always_ok <- "map(x, f,\n    extra_arg = 42)"
+
+  expect_lint(code_hybrid, NULL, indentation_linter())
+  expect_lint(code_hybrid, "Hanging indent", indentation_linter(use_hybrid_indent = FALSE))
+
+  expect_lint(code_non_hybrid, "Indent", indentation_linter())
+  expect_lint(code_non_hybrid, NULL, indentation_linter(use_hybrid_indent = FALSE))
+
+  expect_lint(code_always_ok, NULL, indentation_linter())
+  expect_lint(code_always_ok, NULL, indentation_linter(use_hybrid_indent = FALSE))
+})
+
+test_that("consecutive same-level lints are suppressed", {
```

Comment:
I think it's a good feature, but also not something we do for other linters (e.g. `no_tab_linter()` is the first that comes to mind as a linter that tends to hit on many lines in a row).

Do you think it's worth maintaining the code to remove consecutive lints? Should we consider adding that logic to the other linters too? Maybe it's an argument we could add to `Lint()` or `flatten_lints()` or something like that?

Reply:
I think this is highly context dependent whether consecutive lints need to be "collapsed" into one. This linter tends to produce massive amounts of lints if e.g. a top-level indentation is wrong, see the tables with / without duplicates on the tested packages. But for `no_tab_linter()` I agree it could also be implemented such that it only lints the first line per file that it finds. OTOH if we ever add auto-fixing, this might be detrimental.

Reply:
Filed #1717